src/svm/interface/svm.c

```diff
@@ -1511,6 +1511,80 @@ PetscErrorCode SVMViewScore(SVM svm,PetscViewer v)
   PetscFunctionReturn(0);
 }
 
+#undef __FUNCT__
+#define __FUNCT__ "SVMSetMatGramian"
+/*@
+  SVMSetMatGramian - Set precomputed Gramian (kernel) matrix.
+
+  Input Parameters:
++ svm - SVM context
+- G - precomputed Gramian matrix
+
+  Level: intermediate
+
+.seealso SVM, SVMGetMatGramian(), SVMLoadMatGramian()
+@*/
+PetscErrorCode SVMSetMatGramian(SVM svm,Mat G)
+{
+
+  PetscFunctionBegin;
+  PetscValidHeaderSpecific(svm,SVM_CLASSID,1);
+  PetscValidHeaderSpecific(G,MAT_CLASSID,2);
+
+  TRY( PetscTryMethod(svm,"SVMSetMatGramian_C",(SVM,Mat),(svm,G)) );
+  PetscFunctionReturn(0);
+}
+
+#undef __FUNCT__
+#define __FUNCT__ "SVMGetMatGramian"
+/*@
+  SVMGetMatGramian - Get pracomputed Gramian (kernel) matrix.
+
+  Input Parameter:
+. svm - SVM context
+
+  Output Parameter:
+. G - precomputed Gramian matrix
+
+  Level: intermediate
+
+.seealso SVM, SVMSetMatGramian()
+@*/
+PetscErrorCode SVMGetMatGramian(SVM svm,Mat *G)
+{
+
+  PetscFunctionBegin;
+  PetscValidHeaderSpecific(svm,SVM_CLASSID,1);
+  PetscValidPointer(G,2);
+
+  TRY( PetscUseMethod(svm,"SVMGetMatGramian_C",(SVM,Mat *),(svm,G)) );
+  PetscFunctionReturn(0);
+}
+
+#undef __FUNCT__
+#define __FUNCT__ "SVMComputeOperator"
+/*@
+  SVMComputeOperator - Computes implicit Hessian matrix associated with underlying QP problem.
+
+  Collective on SVM
+
+  Level: advanced
+
+  Input parameter:
+. svm - SVM context
+
+.seealso SVM, SVMSetMatGramian()
+@*/
+PetscErrorCode SVMComputeOperator(SVM svm)
```

Comment:
This function couple two things together:
1. computes the operator
2. sets it as a Hessian of the underlying QP

It could just return the operator and it would then be set separately using SVMSetOperator. I don't like such coupling and doing things under the hood.

Reply:
Okay, I agree. I thought it was early to generalize this thing. However, I've changed my mind when I saw your suggestion. Accepted.